MM-45981_Desktop: Add Server Screen: Improve Onboarding screens

julmondragon opened the pull request:
#### Summary
When the user has no servers configured in the desktop app, a friendlier ‘add server’ screen will be displayed. 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-45981

#### UX Design
https://www.figma.com/file/7BM8S6BL597zBE83U6H8CK/MM-29159-Webapp-Login-Update?node-id=1023%3A92011

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [x] read and understood our [Contributing Guidelines](https://github.com/mattermost/desktop/blob/master/CONTRIBUTING.md)
- [x] completed [Mattermost Contributor Agreement](https://mattermost.com/contribute/)
- [x] executed `npm run lint:js` for proper code formatting

#### Device Information
This PR was tested on: macOS 11.6.8

#### Screenshots
##### Light mode
<img width="1358" alt="Screen Shot 2022-08-30 at 9 19 08 AM" src="https://user-images.githubusercontent.com/79058848/187462091-9bd0b731-6cf6-406e-bb7a-c8f9233027c6.png">

##### Dark mode
<img width="1360" alt="Screen Shot 2022-08-30 at 9 19 32 AM" src="https://user-images.githubusercontent.com/79058848/187462169-974929e8-95d3-4abb-b66d-4b94bb0c83be.png">

##### Mobile view (< 1200px)
<img width="702" alt="Screen Shot 2022-08-30 at 9 19 56 AM" src="https://user-images.githubusercontent.com/79058848/187462237-5600729f-4940-4150-9ea8-c27877e4bac9.png">

#### Release Note
```release-note
Adding Configure Server Screen when starting the app without servers created.
```

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/desktop/22448  
https://circleci.com/gh/mattermost/desktop/22447  
https://circleci.com/gh/mattermost/desktop/22446

julmondragon commented:
> Successfully building:
> https://circleci.com/gh/mattermost/desktop/22448
> https://circleci.com/gh/mattermost/desktop/22447
> https://circleci.com/gh/mattermost/desktop/22446

build-linux-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-linux-x64.tar.gz](https://output.circle-artifacts.com/output/job/a00ecc8b-4ff8-42df-ac7e-3e7ae74c9d89/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-linux-x64.tar.gz)

build-windows-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-ia32.zip](https://output.circle-artifacts.com/output/job/8223bb82-dfed-4d29-8c7d-189750051176/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-ia32.zip)
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-x64.zip](https://output.circle-artifacts.com/output/job/8223bb82-dfed-4d29-8c7d-189750051176/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-x64.zip)

build-mac-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-m1.dmg](https://output.circle-artifacts.com/output/job/38be6ae2-1a3a-4b25-9ec6-46dc1bed1cc9/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-m1.dmg)
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-x64.dmg](https://output.circle-artifacts.com/output/job/38be6ae2-1a3a-4b25-9ec6-46dc1bed1cc9/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-x64.dmg)

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/desktop/22457  
https://circleci.com/gh/mattermost/desktop/22456  
https://circleci.com/gh/mattermost/desktop/22455

julmondragon commented:
> Successfully building:
> https://circleci.com/gh/mattermost/desktop/22457
> https://circleci.com/gh/mattermost/desktop/22456
> https://circleci.com/gh/mattermost/desktop/22455

build-linux-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-linux-x64.tar.gz](https://output.circle-artifacts.com/output/job/0c2fd21a-a49f-4db3-8c2a-d4f12294a403/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-linux-x64.tar.gz)

build-windows-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-ia32.zip](https://output.circle-artifacts.com/output/job/26e77292-58e4-409d-9c62-be53675a6fab/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-ia32.zip)
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-x64.zip](https://output.circle-artifacts.com/output/job/26e77292-58e4-409d-9c62-be53675a6fab/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-x64.zip)

build-mac-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-m1.dmg](https://output.circle-artifacts.com/output/job/d8a9a0bd-176c-456a-a1b9-0a0af70eabe4/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-m1.dmg)
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-x64.dmg](https://output.circle-artifacts.com/output/job/d8a9a0bd-176c-456a-a1b9-0a0af70eabe4/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-x64.dmg)

devinbinnie commented on src/main/preload/modalPreload.js:
```diff
@@ -75,6 +77,9 @@ window.addEventListener('message', async (event) => {
     case GET_LANGUAGE_INFORMATION:
         window.postMessage({type: RETRIEVED_LANGUAGE_INFORMATION, data: await ipcRenderer.invoke(GET_LANGUAGE_INFORMATION)});
         break;
+    case URL_OPEN_EXTERNAL:
```
I don't think doing a message pass here makes a lot of sense since we can handle links the same way we do the Mattermost views. I've added a PR that will allow us to just have a regular link in the modal that should behave the same way that the Mattermost view links do: https://github.com/mattermost/desktop/pull/2247

julmondragon replied:
Updated to use regular link based on https://github.com/mattermost/desktop/pull/2247

devinbinnie commented on src/renderer/components/SaveButton/index.ts:
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+export {default} from './SaveButton';
```
Doesn't seem like this is necessary, we should be able to just directly import this.

julmondragon replied:
Done

devinbinnie commented on src/renderer/components/Input/index.ts:
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+export {default, STATUS, SIZE} from './Input';
```
Same here.

julmondragon replied:
Done

devinbinnie commented on src/renderer/components/ConfigureServer/index.ts:
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+export {default} from './ConfigureServer';
```
And here.

julmondragon replied:
Done

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/desktop/22528  
https://circleci.com/gh/mattermost/desktop/22527  
https://circleci.com/gh/mattermost/desktop/22526

julmondragon commented:
> Successfully building:
> https://circleci.com/gh/mattermost/desktop/22528
> https://circleci.com/gh/mattermost/desktop/22527
> https://circleci.com/gh/mattermost/desktop/22526

build-windows-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-ia32.zip](https://output.circle-artifacts.com/output/job/ab68ca9d-e63b-4251-a794-9dacc123ce78/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-ia32.zip)
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-x64.zip](https://output.circle-artifacts.com/output/job/ab68ca9d-e63b-4251-a794-9dacc123ce78/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-x64.zip)

build-linux-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-linux-x64.tar.gz](https://output.circle-artifacts.com/output/job/9140efa9-dd98-42ab-83ca-df131b5bc588/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-linux-x64.tar.gz)

build-mac-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-m1.dmg](https://output.circle-artifacts.com/output/job/5c7a0cba-b778-45ed-84fe-a23eeacdb643/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-m1.dmg)
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-x64.dmg](https://output.circle-artifacts.com/output/job/5c7a0cba-b778-45ed-84fe-a23eeacdb643/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-x64.dmg)

julmondragon commented on src/renderer/components/ConfigureServer/ConfigureServer.tsx:
```diff
@@ -0,0 +1,354 @@
+
+                    window.removeEventListener('message', handler);
+                    console.log(`newURL: ${newURL}`);
+                    resolve(newURL);
```
Nice catch, updated!

matthewbirtch reviewed (changes requested):
Thanks for this @julmondragon. Looking good. Just a few things I noticed:

1. I'm getting a white flash in dark mode when transitioning between the welcome and the server screen. This same transition happens on light mode, but it is much less noticeable because the two screens already have very light backgrounds. Is it possible to adjust the background color in dark mode so it's not so jarring?
![2022-09-06 11 31 37](https://user-images.githubusercontent.com/2040554/188676033-0108ebfb-28a4-4315-9b83-84cac56386c8.gif)

2. In dark mode, the 'Connect' button's text label is the wrong color when disabled. Can we change it to the white text color and reduce it's opacity (32% I think)
<img width="464" alt="image" src="https://user-images.githubusercontent.com/2040554/188678119-2eb622a0-b2f2-4e56-b798-5a005d4ebc1c.png">

3. Can we remove the `i` icons next to the help text that shows under the fields? It's unneccesary and a bit distracting I think. These weren't in the original design either

4. That layout overall seems like it's sitting too low to my eyes. Can we make it so this area is vertically centered within the view? Let me know what you think.
![image](https://user-images.githubusercontent.com/2040554/188681453-ecfbf92a-38cc-4296-ac56-bf51cbdae3d4.png)

devinbinnie commented on src/renderer/modals/configureServer/configureServer.tsx:
```diff
@@ -0,0 +1,83 @@
+        case DARK_MODE_CHANGE:
+            setDarkMode(event.data.data as boolean);
+            break;
+        case RESIZE_MODAL: {
```
Is there any reason we can't just do this with a `window.on('resize')` hook instead? I'd like to avoid any additional plumbing that doesn't require specific data from the main process that can't be found elsewhere.

julmondragon replied:
Ack'd. Updated

devinbinnie commented on src/renderer/components/ConfigureServer.tsx:
```diff
@@ -0,0 +1,355 @@
+        }
+
+        if (!urlUtils.isValidURL(fullURL)) {
+            formatMessage({
```
I think this is missing a `return` as per my other comment.

julmondragon replied:
Good catch, added.

matthewbirtch commented:
> 2\. Also for @matthewbirtch, we cannot have a link to our Cloud hosting page in our app on the Mac App Store. It violates Apple's restrictions on advertising purchases outside of the application, and they will reject us. We can leave it there if we really want for the standalone app, but it must be disabled for the MAS build.

@devinbinnie @julmondragon I guess we have to remove this then. I don't think we have another page to send people to on mattermost.com that is more informational about creating a cloud server.

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/desktop/22660  
https://circleci.com/gh/mattermost/desktop/22659  
https://circleci.com/gh/mattermost/desktop/22658

julmondragon commented:
> Successfully building:
> https://circleci.com/gh/mattermost/desktop/22660
> https://circleci.com/gh/mattermost/desktop/22659
> https://circleci.com/gh/mattermost/desktop/22658

build-linux-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-linux-x64.tar.gz](https://output.circle-artifacts.com/output/job/d1d70558-91ac-44db-9067-6bf57b084c5e/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-linux-x64.tar.gz)

build-windows-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-ia32.zip](https://output.circle-artifacts.com/output/job/cd367eb7-00f0-4171-9ac4-88f491240988/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-ia32.zip)
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-x64.zip](https://output.circle-artifacts.com/output/job/cd367eb7-00f0-4171-9ac4-88f491240988/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-x64.zip)

build-mac-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-m1.dmg](https://output.circle-artifacts.com/output/job/08b33d46-54be-40c6-8fbf-556a5f3d5f46/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-m1.dmg)
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-x64.dmg](https://output.circle-artifacts.com/output/job/08b33d46-54be-40c6-8fbf-556a5f3d5f46/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-x64.dmg)

julmondragon commented:
@matthewbirtch :
1. Removed the white flash between welcome and configure screens.
2. Updated the color for Connect button disabled
3. Removed the icons for Input's info messages. This was just replicating the design from webapp component.
4. The alternate link was causing the content to look a little bit pushed to the bottom, now that it was removed all is centered.

julmondragon commented on src/renderer/modals/configureServer/configureServer.tsx:
```diff
@@ -0,0 +1,83 @@
+    );
+};
+
+const start = async () => {
```
All modals have it so just followed the pattern. It doesn't seem to have any effect so removed it from welcome modal.

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/desktop/22692  
https://circleci.com/gh/mattermost/desktop/22691  
https://circleci.com/gh/mattermost/desktop/22690

julmondragon commented:
> Successfully building:
> https://circleci.com/gh/mattermost/desktop/22692
> https://circleci.com/gh/mattermost/desktop/22691
> https://circleci.com/gh/mattermost/desktop/22690

build-linux-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-linux-x64.tar.gz](https://output.circle-artifacts.com/output/job/b1449a48-7820-4962-b88d-c555bcc6a26b/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-linux-x64.tar.gz)

build-windows-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-ia32.zip](https://output.circle-artifacts.com/output/job/2c640953-8bbe-488d-b1db-4bba75aa9a5c/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-ia32.zip)
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-x64.zip](https://output.circle-artifacts.com/output/job/2c640953-8bbe-488d-b1db-4bba75aa9a5c/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-win-x64.zip)

build-mac-pr:
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-m1.dmg](https://output.circle-artifacts.com/output/job/4bbff57f-03cc-476c-aca2-0c833bfb4a84/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-m1.dmg)
[tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-x64.dmg](https://output.circle-artifacts.com/output/job/4bbff57f-03cc-476c-aca2-0c833bfb4a84/artifacts/0/tmp/artifacts/mattermost-desktop-5.2.0-develop.1-mac-x64.dmg)

julmondragon commented:
@matthewbirtch @devinbinnie All remaining feedback is addressed now. Thank you!

matthewbirtch reviewed (approved):
Looks great @julmondragon! Thanks

devinbinnie commented:
> 1. There is no immediate validation on the server URL box to check if the URL is even valid to begin with, we can do that using `urlUtils.isValidUrl`. Check the `NewServerModal.tsx` code to see how we do it there.

@julmondragon This bit is still missing, you can still type in whatever you want and hit Connect before validation occurs.

devinbinnie commented:
I also noticed that if you remove all your servers, even after you've seen the Welcome screen you'll see it a second time.
We should only be showing the Welcome screen once, but we can show the new modal.

devinbinnie reviewed (changes requested):
Code looks good, but we should address the issues I commented above.

julmondragon commented:
> > 1. There is no immediate validation on the server URL box to check if the URL is even valid to begin with, we can do that using `urlUtils.isValidUrl`. Check the `NewServerModal.tsx` code to see how we do it there.
> 
> @julmondragon This bit is still missing, you can still type in whatever you want and hit Connect before validation occurs.

This follows the pattern used in the webapp login/signup pages, where fields' validations occurs when executing the action. The only live validation is for empty fields to activate the action button. 
cc @matthewbirtch

julmondragon commented:
> I also noticed that if you remove all your servers, even after you've seen the Welcome screen you'll see it a second time. We should only be showing the Welcome screen once, but we can show the new modal.

This was a requirement from UX implemented as part of the first stage of the onboarding work.
https://github.com/mattermost/desktop/pull/2220#issuecomment-1209475069

cc @matthewbirtch

matthewbirtch commented:
> > I also noticed that if you remove all your servers, even after you've seen the Welcome screen you'll see it a second time. We should only be showing the Welcome screen once, but we can show the new modal.
> 
> This was a requirement from UX implemented as part of the first stage of the onboarding work. [#2220 (comment)](https://github.com/mattermost/desktop/pull/2220#issuecomment-1209475069)
> 
> cc @matthewbirtch

Yep, this is as designed and intended

matthewbirtch commented:
> @matthewbirtch I was under the impression that we should only show the Welcome modal once, since it's just an onboarding screen and if a user removes their server they don't really need to see it again. I think this is something we should consider since it doesn't seem like a good experience to show it multiple times, to me at least.

We discussed it as a team during the walkthrough and agreed to bring back the welcome screen when you've removed all your servers because it acts as a good 'reset'.

I don't see the vast majority of users removing all of their servers multiple times - if ever. If they do happen to do this, I don't see a big problem bringing them back to the welcome screen to orient them again with the product.

Is this a blocker for you @devinbinnie?

devinbinnie commented:
> > @matthewbirtch I was under the impression that we should only show the Welcome modal once, since it's just an onboarding screen and if a user removes their server they don't really need to see it again. I think this is something we should consider since it doesn't seem like a good experience to show it multiple times, to me at least.
> 
> We discussed it as a team during the walkthrough and agreed to bring back the welcome screen when you've removed all your servers because it acts as a good 'reset'.
> 
> I don't see the vast majority of users removing all of their servers multiple times - if ever. If they do happen to do this, I don't see a big problem bringing them back to the welcome screen to orient them again with the product.
> 
> Is this a blocker for you @devinbinnie?

I definitely see the value in coming back to the new Add Server modal added in this PR, since it's very welcoming and it shows the user that they will need to put in another server in order to use the app. Showing the Welcome screen a second time just adds extra clicks and shows them information that they've already seen.

It's not a blocker for me, but I think we should revisit this in a follow-up ticket if not right now. 2/5 on doing this.

matthewbirtch commented:
> I definitely see the value in coming back to the new Add Server modal added in this PR, since it's very welcoming and it shows the user that they will need to put in another server in order to use the app. Showing the Welcome screen a second time just adds extra clicks and shows them information that they've already seen.
> 
> It's not a blocker for me, but I think we should revisit this in a follow-up ticket if not right now. 2/5 on doing this.

Appreciate your thoughts here @devinbinnie. Quite honestly, I don't feel that strongly either, but as mentioned I do think that the welcome screens are a good reset and an opportunity to show some visual polish and interest in an empty state. It's also an opportunity to re-introduce users to the platform. I don't see that as a bad thing. If this was a state that I thought people would find themselves in very often, I might think differently, but I don't think that will be the case.

Again, I don't feel that strongly, so I'm open to change, but I think it's ready to merge as is and if we get feedback from users regularly being annoyed by this, we can address it. I'm perfectly fine with being proven wrong if users provide this feedback!

> As for the URL validation, we can do that in a follow-up ticket as well if need be. Let me know how you'd like to proceed.

I agree that the live URL validation is a nice-to-have and could be looked at later given that we don't seem to have it in the current app anyway so there isn't a regression introduced here.
